crates/store/src/db/models/queries/accounts.rs

```diff
@@ -1010,3 +1205,235 @@ pub(crate) struct AccountStorageMapRowInsert {
+    let raw: Vec<(Vec<u8>, Option<Vec<u8>>)> = SelectDsl::select(
+        t::table
+            .filter(t::account_id.eq(&account_id_bytes))
+            .filter(t::block_num.le(block_num_sql))
+            .order(t::block_num.desc())
+            .limit(1),
+        (t::vault_key, t::asset),
+    )
+    .load(conn)?;
+
+    let entries = raw
+        .into_iter()
+        .filter_map(|(key_bytes, maybe_asset_bytes)| {
+            let key = Word::read_from_bytes(&key_bytes).ok()?;
+            let asset_bytes = maybe_asset_bytes?;
+            let value = Word::read_from_bytes(&asset_bytes).ok()?;
+            Some((key, value))
+        })
+        .collect();
+
+    Ok(entries)
+}
+
+/// Computes the storage commitment from a list of slot commitments.
+///
+/// This replicates the logic from `AccountStorage::commitment()` which hashes all slot
+/// commitments together.
+///
+/// # Arguments
+///
+/// * `slot_commitments` - Vector of slot commitment words
+///
+/// # Returns
+///
+/// The storage commitment as a `Word`
+fn compute_storage_commitment(slot_commitments: &[Word]) -> Word {
+    use miden_objects::crypto::hash::rpo::Rpo256;
+
+    let elements: Vec<Felt> = slot_commitments.iter().flat_map(|w| w.iter()).copied().collect();
+
+    Rpo256::hash_elements(&elements)
+}
+
+/// Helper function to check if a block exists in the `block_headers` table.
+///
+/// This should be called by all `_at_block` query functions to ensure that
+/// queries are only performed against blocks that have been produced.
+///
+/// # Arguments
+///
+/// * `conn` - Database connection
+/// * `block_num` - The block number to check
+///
+/// # Returns
+///
+/// * `Ok(())` - If the block exists
+/// * `Err(DatabaseError::BlockNotFound)` - If the block doesn't exist
+/// * `Err(DatabaseError)` - If there's a database error
+fn block_exists(conn: &mut SqliteConnection, block_num: BlockNumber) -> Result<(), DatabaseError> {
+    use schema::block_headers;
+
+    let count: i64 = SelectDsl::select(
+        block_headers::table.filter(block_headers::block_num.eq(block_num.to_raw_sql())),
+        diesel::dsl::count(block_headers::block_num),
+    )
+    .first(conn)?;
+
+    if count > 0 {
+        Ok(())
+    } else {
+        Err(DatabaseError::BlockNotFound(block_num))
+    }
+}
+
+/// Queries the account code for a specific account at a specific block number.
+///
+/// Returns `None` if:
+/// - The account doesn't exist at that block
+/// - The account has no code (private account or account without code commitment)
+///
+/// # Arguments
+///
+/// * `conn` - Database connection
+/// * `account_id` - The account ID to query
+/// * `block_num` - The block number at which to query the account code
+///
+/// # Returns
+///
+/// * `Ok(Some(Vec<u8>))` - The account code bytes if found
+/// * `Ok(None)` - If account doesn't exist or has no code
+/// * `Err(DatabaseError)` - If there's a database error
+pub(crate) fn select_account_code_at_block(
```

Comment:
Would be awesome to add SQL statements describing these queries (as we do for some similar methods).